toolkit/tools/pkg/imagecustomizerlib/imagecustomizer.go

```diff
@@ -71,133 +152,236 @@ func CustomizeImageWithConfigFile(buildDir string, configFile string, imageFile
+		outputImageFormat, outputImageFile)
 	if err != nil {
-		return fmt.Errorf("invalid image config:\n%w", err)
+		return fmt.Errorf("failed to create image customizer parameters object:\n%w", err)
 	}
+	defer func() {
+		cleanupErr := cleanUp(imageCustomizerParameters)
+		if cleanupErr != nil {
+			if err != nil {
+				err = fmt.Errorf("%w:\nfailed to clean-up:\n%w", err, cleanupErr)
+			} else {
+				err = fmt.Errorf("failed to clean-up:\n%w", cleanupErr)
+			}
+		}
+	}()
 
-	// Normalize 'buildDir' path.
-	buildDirAbs, err := filepath.Abs(buildDir)
+	// ensure build and output folders are created up front
+	err = os.MkdirAll(imageCustomizerParameters.buildDirAbs, os.ModePerm)
 	if err != nil {
 		return err
 	}
 
-	// Create 'buildDir' directory.
-	err = os.MkdirAll(buildDirAbs, os.ModePerm)
+	err = os.MkdirAll(imageCustomizerParameters.outputImageDir, os.ModePerm)
 	if err != nil {
 		return err
 	}
 
-	// Convert image file to raw format, so that a kernel loop device can be used to make changes to the image.
-	rawImageFile := filepath.Join(buildDirAbs, BaseImageName)
+	inputIsoArtifacts, err := convertInputImageToWriteableFormat(imageCustomizerParameters)
+	if err != nil {
+		return fmt.Errorf("failed to convert input image to a raw image:\n%w", err)
+	}
 	defer func() {
-		cleanupErr := file.RemoveFileIfExists(rawImageFile)
-		if cleanupErr != nil {
-			if err != nil {
-				err = fmt.Errorf("%w:\nfailed to clean-up (%s): %w", err, rawImageFile, cleanupErr)
-			} else {
-				err = fmt.Errorf("failed to clean-up (%s): %w", rawImageFile, cleanupErr)
+		if inputIsoArtifacts != nil {
+			cleanupErr := inputIsoArtifacts.cleanUp()
+			if cleanupErr != nil {
+				if err != nil {
+					err = fmt.Errorf("%w:\nfailed to clean-up iso builder state:\n%w", err, cleanupErr)
+				} else {
+					err = fmt.Errorf("failed to clean-up iso builder state:\n%w", cleanupErr)
+				}
 			}
 		}
 	}()
 
-	logger.Log.Infof("Creating raw base image: %s", rawImageFile)
-	err = shell.ExecuteLiveWithErr(1, "qemu-img", "convert", "-O", "raw", imageFile, rawImageFile)
+	err = customizeOSContents(imageCustomizerParameters)
+	if err != nil {
+		return fmt.Errorf("failed to customize raw image:\n%w", err)
+	}
+
+	err = convertWriteableFormatToOutputImage(imageCustomizerParameters, inputIsoArtifacts)
 	if err != nil {
-		return fmt.Errorf("failed to convert image file to raw format:\n%w", err)
+		return fmt.Errorf("failed to convert customized raw image to output format:\n%w", err)
+	}
+
+	logger.Log.Infof("Success!")
+
+	return nil
+}
+
+func convertInputImageToWriteableFormat(ic *ImageCustomizerParameters) (*LiveOSIsoBuilder, error) {
+	logger.Log.Infof("Converting input image to a writeable format")
+
+	if ic.inputImageFormat == ".iso" {
+
+		inputIsoArtifacts, err := createIsoBuilderFromIsoImage(ic.buildDir, ic.buildDirAbs, ic.inputImageFile)
+		if err != nil {
+			var cleanUpError error
```

Comment:
This cleanup should occur in `createIsoBuilderFromIsoImage`, not in the caller of `createIsoBuilderFromIsoImage`.

It might be helpful to create a `createIsoBuilderFromIsoImageHelper` function that handles all the initialization logic. Then have the `createIsoBuilderFromIsoImage` call the `createIsoBuilderFromIsoImageHelper` function and call the cleanup logic if `createIsoBuilderFromIsoImageHelper` fails.

Reply:
makes sense. I've moved `isobuilder.cleanUp()` into `createIsoBuilderFromIsoImage()` and set it such that it gets called only if there's an error. This 'only if' avoid having to add the call to each error handling in the rest of `createIsoBuilderFromIsoImage()` (and I think we can do without the helper function).
```
func someFunc() (err error ) {
    x := &SomeType{}
    defer func() { if err != nil {x.cleanUp} }()
    err = anotherFunc()
    if err != nil { return err } // -> this invokes the clean-up code.
    return nil // -> this does not invoke the clean-up code.
}